File: paper/test_palettes.py
"""Verify project palettes load correctly and render a preview figure."""

from __future__ import annotations

import os
from pathlib import Path

import matplotlib

matplotlib.use("Agg")
import matplotlib.pyplot as plt

from palettes import get_palette, load_palettes

OUTDIR = Path(__file__).resolve().parent / "figures"
os.makedirs(OUTDIR, exist_ok=True)


def plot_palette_preview() -> None:
    """Render all palettes as horizontal color bars."""
    pals = load_palettes()
    names = [k for k in pals.keys() if not k.startswith("_")]
    n_pals = len(names)

    fig, ax = plt.subplots(figsize=(10, 0.6 * n_pals + 1))
    ax.set_xlim(0, 10)
    ax.set_ylim(0, n_pals)
    ax.axis("off")

    for i, name in enumerate(names):
        colors = get_palette(name)
        n_cols = len(colors)
        y = n_pals - i - 1
        for j, col in enumerate(colors):
            x_start = j * (10 / n_cols)
            width = 10 / n_cols
            ax.add_patch(plt.Rectangle((x_start, y), width, 0.8, color=col, ec="white"))
        ax.text(-0.2, y + 0.4, name, ha="right", va="center", fontsize=9)

    ax.set_title("Project Publication Palette Preview", fontsize=12, weight="bold", pad=10)
    png_path = OUTDIR / "palette_preview.png"
    pdf_path = OUTDIR / "palette_preview.pdf"
    plt.tight_layout()
    fig.savefig(png_path, dpi=300, bbox_inches="tight")
    fig.savefig(pdf_path, bbox_inches="tight")
    plt.close(fig)
    print(f"[OK] PNG: {png_path}")
    print(f"[OK] PDF: {pdf_path}")


if __name__ == "__main__":
    print("Loading palettes...")
    names = sorted(k for k in load_palettes().keys() if not k.startswith("_"))
    print("Available:", ", ".join(names))
    for name in names:
        cols = get_palette(name)
        print(f"  {name}: {len(cols)} colors")
    plot_palette_preview()
    print("[DONE] All palettes verified.")

File: paper/test_palettes.R
# Verify project palettes load correctly and render a preview figure

suppressPackageStartupMessages({
  library(ggplot2)
})

source("d:/铁衰老 绝不重蹈覆辙/paper/palettes.R")

outdir <- "d:/铁衰老 绝不重蹈覆辙/paper/figures"
dir.create(outdir, showWarnings = FALSE, recursive = TRUE)

pals <- load_palettes()
names <- setdiff(names(pals), "_meta")

cat("Available palettes:\n")
cat(paste(" ", names, collapse = "\n"), "\n")

# Build a long data frame for ggplot2
df_list <- lapply(names, function(nm) {
  cols <- get_palette(nm)
  data.frame(
    palette = nm,
    idx = seq_along(cols),
    color = cols,
    stringsAsFactors = FALSE
  )
})
df <- do.call(rbind, df_list)
df$palette <- factor(df$palette, levels = names)

p <- ggplot(df, aes(x = idx, y = palette, fill = color)) +
  geom_tile(color = "white", linewidth = 0.3) +
  scale_fill_identity() +
  labs(title = "Project Publication Palette Preview",
       x = NULL, y = NULL) +
  theme_minimal(base_size = 10) +
  theme(
    plot.title = element_text(face = "bold", size = 12),
    axis.text.x = element_blank(),
    axis.ticks = element_blank(),
    panel.grid = element_blank()
  )

png_path <- file.path(outdir, "palette_preview_r.png")
pdf_path <- file.path(outdir, "palette_preview_r.pdf")

ggsave(png_path, p, width = 10, height = 0.6 * length(names) + 1, dpi = 300, units = "in")
ggsave(pdf_path, p, width = 10, height = 0.6 * length(names) + 1, units = "in")

cat(sprintf("[OK] PNG: %s\n", png_path))
cat(sprintf("[OK] PDF: %s\n", pdf_path))
cat("[DONE] All palettes verified.\n")
